### ISSUE TITLE
designspaceLib.newDefaultLocation must also make non-mapped location

### ISSUE THREAD

**Issue by LettError** [fonttools/fonttools#1655]
`designspaceLib.newDefaultLocation` must be able to generate a location with the actual axis defaults, with "designspace", non-mapped values. I will also review the other uses of `axisDescriptor.map_forward`.

https://github.com/fonttools/fonttools/blob/18f969ab7dca9b5ec16dcf480ffe3c46ae70beb7/Lib/fontTools/designspaceLib/__init__.py#L1128

**Comment by anthrotype** [comment]
Hey Erik,

> the actual axis defaults, with "designspace", non-mapped values

well, that's what it _is_ doing.
Mapping "forward" in this context means from user to internal design coordinates.
The first column of the axes maps is "user" (the 'input'); the second column (the 'output') is internal "design" coordinates.
Also -- and that may be the source of the confusion here -- the axes minimum, default and maximum are always specified as "user"-space coordinates. That is why, when you ask "give me a new default location, in design coordinates", you first take the axis.default (always in 'user' coordinates) and map that forward to 'design' coordinate space.

**Comment by anthrotype** [comment]
See also my other comment in https://github.com/googlefonts/fontmake/pull/552#issuecomment-497431705

**Comment by LettError** [comment]
>  the axes minimum, default and maximum are always specified as "user"-space coordinates.
is it really? that's not how I built it originally and I'm not sure that is a good idea.

**Comment by anthrotype** [comment]
yes, that's how varLib interprets them...

Quoting @behdad from https://github.com/LettError/designSpaceDocument/issues/16

> One thing that still bothers me is that the axis default / minimum / maximum values are in the user space, whereas everything else is by default in the design space.

and then, in another later comment, he added

> I think axis default/min/max should default to "input" space. Ie. they define the user-visible range, and the mapping defines how to convert that to internal representation. For masters and instances allowing both would be great.

**Comment by LettError** [comment]
That is not always convenient.

**Comment by madig** [comment]
(FWIW, I stumbled over this, too, and added clarifications to https://github.com/fonttools/fonttools/tree/master/Doc/source/designspaceLib#axisdescriptor-object)

**Comment by LettError** [comment]
Great I think google is happy then.

**Comment by anthrotype** [comment]
> That is not always convenient. And there is a lot of data that does not work like that.

this is not good, and I do share your concern. We need to come up together with a reasonable plan to make so that these location coordinates are set explicitly and unambiguously as either user-space or design-space.

I'd also like @behdad to comment on this issue.

**Comment by madig** [comment]
glyphsLib and some other things rely on axis default/min/max to always be in user space. Which code assumes it is always design space? Not sure this can be solved without breaking one or the other. I am however open to breakage if we get something sensible in.

Here's a sketch:

```Python
import fontTools.designspaceLib


class AxisDescriptor(fontTools.designspaceLib.AxisDescriptor):
    # default: int  # design space
    # minimum: int  # design space
    # maximum: int  # design space

    @property
    def default_user(self):
        return self.map_backward(self.default)

    @default_user.setter
    def default_user(self, value):
        self.default = self.map_forward(value)

    @property
    def minimum_user(self):
        return self.map_backward(self.minimum)

    @minimum_user.setter
    def minimum_user(self, value):
        self.minimum = self.map_forward(value)

    @property
    def maximum_user(self):
        return self.map_backward(self.maximum)
    
    @maximum_user.setter
    def maximum_user(self, value):
        self.maximum = self.map_forward(value)
```

The idea is to always keep the axis values in design space and have client code ask for user coordinates explicitly. I'm not sure the setters are useful as written here, because:

```
a = AxisDescriptor()
a.minimum = 1
a.default = 5
a.maximum = 10
a.map = ((100, 1), (400, 5), (900, 10))

a.default_user = 405
-> a.default == 5.05
```

So I moved the default location. Not sure this is useful. Maybe no setters for user-locations? Or make them modify the mapping instead? ... does that make more or less sense...

**Comment by LettError** [comment]
> Which code assumes it is always design space?
All the designspace consumers that were written before these modifications and the projects that were built with them. I'll have to mop up.

**Comment by LettError** [comment]
I find it ugly that now you need to do math to find out which master is the default, where it was perfectly clear before. What was the motivation to make the axis values in userspace coordinates?

**Comment by behdad** [comment]
> > Which code assumes it is always design space?
> 
> All the designspace consumers that were written before these modifications and the projects that were built with them. I'll have to mop up.

Before there was the map element there was no distinction before the two.  So I don't think that a lot of prior code / projects are involved.

**Comment by behdad** [comment]
> I find it ugly that now you need to do math to find out which master is the default, where it was perfectly clear before.

There's no math involved.  Just a dictionary lookup really.

> What was the motivation to make the axis values in userspace coordinates?

It made more sense to me.  Since the axis default/min/max are highly user-visible values, I think it makes sense that they are declared directly in that space, whereas the design space is internal working of the font sources.

At any rate, let's pick a clear path forward and implement it.  Not making decisions just keeps the problem lingering forward.

One can imagine inferring min/default/max from map element itself.  min/max are clear.  One would need to add an attribute for default...  That makes the designspace file unambiguously clear.

**Comment by belluzj** [comment]
Hello! I don't have an opinion on whether those values should be in user or design space, but I wanted to say that if you're ready to change the name of those parameters, I think it would be useful to add "design" or "user" in their name. That would be the definitive way to clarify the situation, even in day-to-day use by designers, not only for tool developers. E.g. `userMin=300 userDefault=400 userMax=800`

**Comment by LettError** [comment]
Is it required for variable fonts that the axis default value has an entry in the axis map?

**Comment by anthrotype** [comment]
> Is it required for variable fonts that the axis default value has an entry in the axis map?

Yes, otherwise you may get this issue https://github.com/fonttools/fonttools/issues/1631

**Comment by LettError** [comment]
That should be in the designspace spec then.
